Conducting an A/B test involves comparing two or more groups to assess the impact of a change or intervention. In our case, let's define a clear objective and hypothesis for the A/B test using logistic dataset. Since I've data with order details and freight rates, I can formulate an objective related to optimizing transportation costs.

## Objective:

To determine whether a change in freight rates for two different carrier type significantly impacts transportation costs for order deliveries.

## Steps to Implement:

Identify orders with the A carrier type and the B carrier type.

Calculate transportation costs for each order based on the corresponding freight rates.

Split the dataset into control and experimental groups.

Conduct the A/B test using statistical analysis.

## Importing Libraries :

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sqlalchemy
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings('ignore')
from scipy.stats import t
from scipy.stats import ttest_ind

## Import data from database :

In [2]:
engine = sqlalchemy.create_engine('postgresql://postgres:4545@localhost/analysis_in_python')

query_1 = "SELECT * FROM logistic_analysis.order_details;"

query_2 = "SELECT * FROM logistic_analysis.freight_rates;"

order_df = pd.read_sql(query_1, engine)

freight_df = pd.read_sql(query_2, engine) 

engine.dispose()

In [3]:
order_df.head()

order_id  order_date origin_port carrier  tpt service_level  \
0  1.447296e+09  2014-08-13      PORT09   V44_3    1           CRF   
1  1.447158e+09  2014-05-06      PORT09   V44_3    1           CRF   
2  1.447139e+09  2013-07-14      PORT09   V44_3    1           CRF   
3  1.447364e+09  2014-04-04      PORT09   V44_3    1           CRF   
4  1.447364e+09  2014-05-10      PORT09   V44_3    1           CRF   

   ship_ahead_day_count  ship_late_day_count   customer  product_id  \
0                     3                    0  V55555_53     1700106   
1                     3                    0  V55555_53     1700106   
2                     3                    0  V55555_53     1700106   
3                     3                    0  V55555_53     1700106   
4                     3                    0  V55555_53     1700106   

  plant_code destination_port  unit_quantity  weight  
0    PLANT16           PORT09            808   14.30  
1    PLANT16           PORT09           3188   87.94  
2    PLANT16           PORT09           2331   61.20  
3    PLANT16           PORT09            847   16.16  
4    PLANT16           PORT09           2163   52.34

In [4]:
freight_df.head()

carrier origin_portcd dest_portcd  min_weight_qty  max_weight_qty svccd  \
0  V444_4        PORT04      PORT09           13.01           13.50   DTD   
1  V444_8        PORT09      PORT09          100.00          249.99   DTD   
2  V444_5        PORT02      PORT09          500.00         1999.99   DTP   
3  V444_6        PORT08      PORT09           10.00           14.99   DTD   
4  V444_1        PORT05      PORT09           71.00           99.99   DTD   

   minimum_cost  rate mode_desc  tpt_day_cnt  carrier_type  
0          3.04  1.04    AIR               1  V888888883_1  
1         27.30  0.06    AIR               0   V88888888_0  
2          5.03  0.06    AIR               2   V88888888_0  
3         43.23  1.83    AIR               2   V88888888_0  
4         17.13  0.14    AIR               2  V888888883_1

In [5]:
# Merge datasets based on common columns (e.g., Carrier)

df = pd.merge(order_df, freight_df, how='inner', on=['carrier'])

In [6]:
df.head()

order_id  order_date origin_port carrier  tpt service_level  \
0  1.447385e+09  2014-12-08      PORT09  V444_0    0           DTP   
1  1.447385e+09  2014-12-08      PORT09  V444_0    0           DTP   
2  1.447385e+09  2014-12-08      PORT09  V444_0    0           DTP   
3  1.447385e+09  2014-12-08      PORT09  V444_0    0           DTP   
4  1.447385e+09  2014-12-08      PORT09  V444_0    0           DTP   

   ship_ahead_day_count  ship_late_day_count  customer  product_id plant_code  \
0                     0                    0  V55555_4     1692724    PLANT16   
1                     0                    0  V55555_4     1692724    PLANT16   
2                     0                    0  V55555_4     1692724    PLANT16   
3                     0                    0  V55555_4     1692724    PLANT16   
4                     0                    0  V55555_4     1692724    PLANT16   

  destination_port  unit_quantity    weight origin_portcd dest_portcd  \
0           PORT09            332  1.023143        PORT04      PORT09   
1           PORT09            332  1.023143        PORT03      PORT09   
2           PORT09            332  1.023143        PORT04      PORT09   
3           PORT09            332  1.023143        PORT04      PORT09   
4           PORT09            332  1.023143        PORT04      PORT09   

   min_weight_qty  max_weight_qty svccd  minimum_cost  rate mode_desc  \
0             0.0           99.99   DTD          3.46  0.08    AIR      
1             0.0         5000.00   DTD         27.20  9.20    GROUND   
2           500.0         1999.99   DTD          3.46  0.05    AIR      
3             0.0           99.99   DTP          1.50  0.05    AIR      
4           100.0          249.99   DTP          1.50  0.04    AIR      

   tpt_day_cnt carrier_type  
0            2  V88888888_0  
1            1  V88888888_0  
2            2  V88888888_0  
3            3  V88888888_0  
4            3  V88888888_0

In [7]:
df['carrier'].unique()

array(['V444_0', 'V444_1'], dtype=object)

## Hypothesis:

Null Hypothesis (H0): There is no significant difference in transportation costs between A & B carrier types.

Alternative Hypothesis (H1): There is a significant difference in transportation costs between these carrier types.

In [8]:
# Define the control (A carrier type) and experimental (B carrier type) groups

control_group = df[df['carrier'] == 'V444_0'] # A carrier type

experimental_group = df[df['carrier'] == 'V444_1'] # B carrier type

In [9]:
# control_group['minimum_cost'].plot(kind= 'kde')

# plt.title('KDE - Plot Of Control Group')
# plt.show()

In [10]:
# experimental_group['minimum_cost'].plot(kind= 'kde')

# plt.title('KDE - Plot Of experimental_group')
# plt.show()

In [11]:
# Calculate transportation costs for each group
control_costs = control_group['unit_quantity'] * control_group['minimum_cost']

experimental_costs = experimental_group['unit_quantity'] * experimental_group['minimum_cost']

In [12]:
# Perform t-test to compare means
t_stat, p_value = ttest_ind(control_costs, experimental_costs)

# Set significance level
alpha = 0.05

In [13]:
# Compare p-value with significance level
if p_value < alpha:
    print(f"Reject the null hypothesis. There is a significant difference in transportation costs.")
else:
    print("Fail to reject the null hypothesis. There is no significant difference in transportation costs.")

Reject the null hypothesis. There is a significant difference in transportation costs.


## Interpreting A/B test result : 

### Throughout A/B test I have conclude that their is a signifacnt difference between two carrier. But which one is better ?

__To make a decision based on the result of our A/B test, it will good to consider these factors:__

__Evaluate Effect Size:__

Calculate and examine the effect size (e.g., Cohen's d) to understand the magnitude of the difference between the two carrier types. This will provide insight into the practical significance of the observed effect.

__Consider Business Goals:__

Align the results with the business goals and objectives. Identify which carrier type aligns better with the overall goals of the logistics and supply chain operations.

__Cost-Benefit Analysis:__

Conduct a cost-benefit analysis to assess the impact of the cost difference on overall business. Consider not only the financial aspects but also other qualitative factors like reliability, customer satisfaction, and delivery times.

__Involve Stakeholders:__

Engage relevant stakeholders, including those from logistics, finance, and customer service. Gather input from different perspectives to ensure a comprehensive understanding of the implications of choosing one carrier over the other.

__Explore Long-Term Impact:__

Consider the long-term implications of selecting a specific carrier. Some carriers may offer discounts for long-term partnerships, and a stable and reliable carrier relationship can contribute to the success of supply chain.

__Risk Assessment:__

Assess any risks associated with each carrier type. Evaluate factors such as the carrier's track record, potential disruptions, and the resilience of their operations.

__Decision Framework:__

Establish a decision framework that incorporates both quantitative and qualitative factors. Assign weights to different criteria based on their importance to the business, and use a systematic approach to evaluate the options.

__Consult Data Trends:__

Examine trends in historical data related to the carriers. Look for patterns or fluctuations that might impact decision. For example, a carrier's performance during peak seasons or specific months could be crucial.

__Feedback Loop:__

Establish a feedback loop to monitor the ongoing performance of the selected carrier. Regularly assess whether the carrier continues to meet expectations and make adjustments as needed.

__Documentation:__

Document the decision-making process and the rationale behind our choice. This documentation can serve as a reference for future decisions and audits.

## I'm considering Cohen's d for now :

Cohen's d is a measure of the effect size, indicating the standardized difference between two means. We can calculate Cohen's d using the following formula:

$
d = \frac{\bar{X}_1 - \bar{X}_2}{s}
$

Where:

- X̄₁ & X̄₂ are the means of the two groups.
- s is the pooled standard deviation.

In [14]:
def cohen_d(group1, group2):
    n1, n2 = len(group1), len(group2)
    mean1, mean2 = np.mean(group1), np.mean(group2)
    var1, var2 = np.var(group1, ddof=1), np.var(group2, ddof=1)
    pooled_var = ((n1 - 1) * var1 + (n2 - 1) * var2) / (n1 + n2 - 2)
    
    d = (mean1 - mean2) / np.sqrt(pooled_var)
    
    return d

cohen_d_value = cohen_d(experimental_costs, control_costs)
print(f"Cohen's d: {cohen_d_value}")

Cohen's d: -0.1559370380091116


## Interpretation of Cohen's d:

Small effect: 0.2 ≤ d < 0.5

Medium effect: 0.5 ≤ d <0.8

Large effect: d ≥ 0.8

__The negative value indicates the direction of the effect:__ a negative Cohen's d suggests that the experimental group (B carrier type) has a lower mean transportation cost than the control group (A carrier type). The magnitude of Cohen's d (-0.1559) indicates a small effect size.

__Things to note :__ that statistical significance does not necessarily imply practical importance, and decisions should consider both statistical and practical significance. It's always essential to interpret results in the context of the specific problem or domain knowledge.